### PR TITLE
Associations selector

### PR DESCRIPTION
**Pull request by romul**
See details at #109 

Currently it's working with `has_many through`
I haven't checked it with direct many_to_many (ecto 2) yet.

### CODE REVIEW

File: lib/ex_admin/show.ex [hunk 1]
```diff
@@ -236,10 +240,143 @@ defmodule ExAdmin.Show do
...
+
+  *Note:* If you have custom keys in intersection table, please use association_filler/2 to specify them explicit.
+
+  ## Examples
+
+      show post do
+        attributes_table
+
+        panel "Tags" do
+          table_for(post, :post_tags) do
```
Comment (smpallen99) [inline lib/ex_admin/show.ex]: Should this not be `table_for(post.post_tags) do`?

File: lib/ex_admin/show.ex [hunk 2]
```diff
@@ -236,10 +240,143 @@ defmodule ExAdmin.Show do
...
+            column :tag
+          end
+          markup_contents do
+            association_filler post, :tags, autocomplete: true
+          end
+        end
+      end
+  """
```
Comment (smpallen99) [inline lib/ex_admin/show.ex]: Example should include the changes required in the model to get this working.

**Comment by smpallen99** [comment]
Spent  about an hour and still can't get simple Post/Tag example working. Can you provide a simple Post/Tag working project that I can pull down and test with. 

I would prefer that that the pretty_name function on the model be optional. May also want to consider the name `display_name`. Take a look at `ExAdmin.Helpers.get_resource_field/3`. I already have code that determines the the display_name of a field. You might be able to use this code, or get some ideas from it.

I would really prefer if this feature could work without the need of the model's `admin_search_query` function. A default query can be derived from the model struct and the keyword (:name for example). So, perhaps we can:

1. look for function in the model and call it if it exists
2. Look for an option with the name and it use it
3. If neither of these exist, use :name if it exists in the schema
4. Use the first field after :id

I've put a lot of work into ExAdmin to use intelligent defaults where ever possible, with the ability to override the defaults. I would really like to keep that approach moving forward :)

Once I can get an example working for this feature, I'll probably have more comments. 

I think  this is a great feature and I appreciate all the work that you have put into it. 

Steve

**Comment by romul** [comment]
@smpallen99 
Yeah, I understand your point. Although it's really a complex set of trade offs.
Intelligent defaults are good, and on other hand too much magic isn't so good.
The popular objection against ActiveAdmin is "easier to write the admin panel from scratch than to figure out how to configure it".
So I want to make ex_admin easier to configure and with ability to override any action for specified resource. Often explicit is better than implicit. In ideal it should raise compile error if something is not working due configuration and error message should contain detailed way to fix it. So, I prefer raise compilation error if `pretty_name` and `admin_search_query` for association used in `association_filler` are not defined and thereby keep it simple and easy to understand.
I take a look at version of `association_filler` with defaults and I see how difficult will be debug it in non-typical cases. Surely someone will suffer the whole day before find that he/she can easily specify all the parameters explicit. If you know how to determine all real keys instead of these assumptions: https://github.com/romul/ex_admin/blob/associations_selector/lib/ex_admin/show.ex#L272..L278, please fix it.
Atm I think, defaults - good, assumptions about defaults - evil, so I really dislike https://github.com/romul/ex_admin/blob/associations_selector/lib/ex_admin/show.ex#L272..L278 :-)

File: lib/ex_admin/show.ex (hunk 1 above)
Comment (romul) [inline lib/ex_admin/show.ex]: yes, I'll fix it

**Comment by romul** [comment]
@smpallen99 Working example: https://github.com/romul/synergy/commit/858224967d9208aea27714c53ff20238c0141334
Btw maybe would be even better to share not-working examples to understand how to make it simpler.

**Comment by smpallen99** [comment]
@romul I will share where I went wrong. But wanted to see the whole thing in action first. I have synergy working so I have now I understand how is supposed to work. For some reason, I thought there was support for creating the associated resource on the fly, like you would expect for a tagging solution. Probably a good feature for later :)

BTW, I’m not that fond of the name `association_filler` but don’t have a better suggestion yet. I’ll come up with some suggestions soon. 

As for the intelligent defaults discussion, I can see your point about “magic”. I have also read the criticism on ActiveAdmin being too complicated to setup. With that said, I would argue that its easier to setup if you don’t have to add stuff to your model like `pretty_name` to get the feature working, as long as there is clear documentation on how to override the implicit feature.

I agree with you that we don’t want people spending a lot of time troubleshooting an issue with the implicit feature. So, I think error checking and useful exceptions are important. 

Steve

**Comment by romul** [comment]
@smpallen99 I guess, I figured out a good solution:
we could add function `display_name` to https://github.com/smpallen99/ex_admin/blob/master/priv/templates/admin.gen.resource/resource.exs with simple default implementation. To be clear I presume a case than display_name is not match to one field, rather something like "#{prop.name} (#{prop.permalink})", b/c just name is not uniq in the many cases.

> support for creating the associated resource on the fly

In my experience even for tags is better to avoid creating on the fly, b/c of typos. Usually it goes like 
1) manager writes to content-manager: "add 'enhencement' tag to this post"
2) content-manager make copy-paste
3) autocomplete fails and a new tag with the typo is created

**Comment by romul** [comment]
So, the generator would generate something like:
``` elixir
defmodule Synergy.ExAdmin.Property do
  use ExAdmin.Register
  import Ecto.Query

  register_resource Synergy.Property do

  end

  def display_name(property) do
    property.name
  end

  def admin_search_query(keywords) do
    from(r in Synergy.Property, where: ilike(r.name, ^("%#{keywords}%")), order_by: [:name])
  end
end
```

And in this case we could use `display_name` and `admin_search_query` in many places of the admin panel. For example, general search on top of the filter at index page would be cool feature, b/c it could use, for example, fulltext-search features of Postgres.

**Comment by smpallen99** [comment]
That is not what I was thinking. I already have the ability to intelligently "guess" at the display name unless there is a `display_name` function in the model. So, I would like to use the same here so its consistent.

I can live with the generator creating a `admin_search_query`. However, I would like code that creates that query if the function does not exist. This way, people can use the new feature without having to update their resource defn file. Thanks...Steve

**Comment by romul** [comment]
@smpallen99 I'm ok with the ability to intelligently "guess". But I'd like to make `display_name` feature more obvious somehow. For example, I discovered it only after you pointed me directly to `ExAdmin.Helpers.get_resource_field/3` 
Maybe some comments in generated files?
```elixir
defmodule Synergy.ExAdmin.Property do
  use ExAdmin.Register
  import Ecto.Query

  register_resource Synergy.Property do

  end

  # This function would be used to display a property inside admin panel
  # uncomment it to override default behavior.
  # def display_name(property) do
  #   property.name
  # end

  # This function would be used to search properties by keywords
  # uncomment it to override default behavior.
  # def admin_search_query(keywords) do
  #   from(r in Synergy.Property, where: ilike(r.name, ^("%#{keywords}%")), order_by: [:name])
  # end
end
```

**Comment by smpallen99** [comment]
Yes, that is my bad. I added this feature but did not document it. I had to go digging in the code to find it, and even that took me some time :(. Just to clarify a few things:
* I’m ok with the idea of commented out functions. As a matter of fact, I was thinking of putting commended out boilerplate inside the `register_resource` do block showing a default `index`, `show`, `form`, and `query` block. This would not be too difficult to generate at `mix admin.gen.resource` time. Could even have option to include/exclude it. I now I’m always having to refer to the docs when I want to add one of the blocks LOL.
* I’m good with adding the commented `display_name` and `admin_search_query` as you have it listed below. However, we still need to update the docs for this feature. I’ll go ahead and add it to the wiki to start with. Furthermore, when we add the rest of the boilerplate, I will likely include exclusion of these items with the exclude option. But that is an issue for down the road. 

I’m a little swamped today getting ready for an Elixir presentation I’m give at a local meetup. I’ll get back to some testing on your branch tomorrow. 

Steve

**Comment by romul** [comment]
> I was thinking of putting commended out boilerplate inside the `register_resource` do block showing a default `index`, `show`, `form`, and `query` block.

👍 
I also thought about it already. It could be an option like `verbose` (true by default) in generator to enable/disable these comments.


> I’m a little swamped today getting ready for an Elixir presentation I’m give at a local meetup. I’ll get back to some testing on your branch tomorrow.

No problem. I'm also busy in the next couple of days. So I guess we will complete work on this branch at the weekend :-)
Have a nice meetup!

**Comment by romul** [comment]
@smpallen99 Take a look at default implementations for `admin_search_query` and `display_name`: https://github.com/smpallen99/ex_admin/pull/114/commits/44e6b1288d8986dac5dffa58544b6d2cdd94e36a

File: lib/ex_admin/ex_admin.ex [hunk 3]
```diff
@@ -180,6 +180,20 @@ defmodule ExAdmin do
...
 
+
+  def get_registered_by_association(resource, assoc_name) do
+    resource_model = resource.__struct__
+    assoc_model = case resource_model.__schema__(:association, assoc_name) do
+      %{through: [link1, link2]} ->
+        resource |> Ecto.build_assoc(link1) |> Ecto.build_assoc(link2) |> Map.get(:__struct__)
+      %{queryable: assoc_model} ->
+        assoc_model
+      nil ->
```
Comment (smpallen99) [inline lib/ex_admin/ex_admin.ex]: Should this match be `_` so it catches everything that does not match on the above clauses?

File: lib/ex_admin/show.ex [hunk 4]
```diff
@@ -308,21 +298,21 @@ defmodule ExAdmin.Show do
...
+              assoc_key: "tag_id", autocomplete: false)
           end
         end
       end
   """
   defmacro association_filler(resource, opts) do
     quote bind_quoted: [resource: resource, opts: opts] do
-      required_opts = [:resource_key, :assoc_name, :assoc_key, :assoc_model]
+      required_opts = [:resource_key, :assoc_name, :assoc_key]
       unless MapSet.subset?(MapSet.new(required_opts), MapSet.new(Dict.keys(opts))) do
```
Comment (smpallen99) [inline lib/ex_admin/show.ex]: Dict has been deprecated. I wonder if we should use `Map.keys` or `Keyword.keys` here (can't remember if opts is a map or a keyword list here...)

**Comment by smpallen99** [comment]
@romul https://github.com/smpallen99/ex_admin/pull/114/commits/44e6b1288d8986dac5dffa58544b6d2cdd94e36a Looks good. Just one change suggested above.

**Comment by smpallen99** [comment]
@romul And of course, will need to fix merge conflicts... Do you think we're ready for a merge after that?

File: lib/ex_admin/ex_admin.ex (hunk 3 above)
Comment (romul) [inline lib/ex_admin/ex_admin.ex]: No, nil clause could be fixed by an user. B/c it means non-existing association.
Other cases would mean breaking changes in Ecto, so in this case user could not fix it by himself.

Comment (romul) [inline lib/ex_admin/ex_admin.ex]: It could be so https://github.com/romul/ex_admin/commit/75eb1367714f74b1c1cbd48812161e2d95784ecd though

Comment (smpallen99) [inline lib/ex_admin/ex_admin.ex]: I see your point about nil. That indicates the association does  not exist. I was worried about not matching nay of the three clauses, and throwing a complicated exception. But perhaps your three match conditions will catch all possible associations. The probably all have a 'queryable' field.

**Comment by romul** [comment]
@smpallen99 
> Do you think we're ready for a merge after that?

Yes, I think it's pretty good for a merge. I have some ideas about further improvements like `many_to_many` support, comments in generated defn files, wider use of display_name, etc., but I'd prefer to have separate issues for these tasks.

**Pull request merged**